Sources/SwiftProtobuf/TextFormatScanner.swift

```diff
@@ -1109,54 +1106,239 @@ internal struct TextFormatScanner {
+                    throw TextFormatDecodingError.unknownField
+                }
+                // Unknown field name
+                break
+            default:
+                if c == terminator {
+                    let _ = skipOptionalObjectEnd(c)
+                    return nil
+                }
+                throw TextFormatDecodingError.malformedText
+            }
+
+            assert(options.ignoreUnknownFields || options.ignoreUnknownExtensionFields)
+            try skipUnknownFieldValue()
+            // Skip any separator before looping around to try for another field.
+            skipOptionalSeparator()
+        }
+    }
+
+    // Helper to skip past an unknown field value, when called `p` will be pointing
+    // at the first character after the unknown field name.
+    internal mutating func skipUnknownFieldValue() throws {
+        // This is modeled after the C++ text_format.cpp `ConsumeField()`
+        //
+        // Guess the type of this field:
+        // - If this field is not a message, there should be a ":" between the
+        //   field name and the field value and also the field value should not
+        //   start with "{" or "<" which indicates the beginning of a message body.
+        // - If there is no ":" or there is a "{" or "<" after ":", this field has
+        //   to be a message or the input is ill-formed.
+
         skipWhitespace()
-        if p == end {
-            return nil
+        if (skipOptionalColon()) {
+            if p == end {
+                // Nothing after the ':'?
+                throw TextFormatDecodingError.malformedText
+            }
+            let c = p[0]
+            if c != asciiOpenAngleBracket && c != asciiOpenCurlyBracket {
+                try skipUnknownPrimativeFieldValue()
+            } else {
+                try skipUnknownMessageFieldValue()
+            }
+        } else {
+            try skipUnknownMessageFieldValue()
+        }
+    }
+
+    /// Helper to see if this could be the start of a hex or octal number so unknown field
+    /// value parsing can decide how to parse/validate.
+    private func isNextNumberFloatingPoint() -> Bool {
+        // NOTE: If we run out of characters can can't tell, say it isn't and let
+        // the other code error handle.
+        var scan = p
+        var c = scan[0]
+
+        // Floats for decimals can have leading '-'
+        if c == asciiMinus {
+            scan += 1
+            if scan == end { return false }
+            c = scan[0]
+        }
+
+        if c == asciiPeriod {
+            return true  // "(-)." : clearly a float
+        }
+
+        if c == asciiZero {
+            scan += 1
+            if scan == end { return false }  // "(-)0[end]" : call it not a float
+            c = scan[0]
+            if c == asciiLowerX ||  // "(-)0x" : hex - not a float
+                (c >= asciiZero && c <= asciiSeven) {  // "(-)0[0-7]" : octal - not a float
```

Comment:
Should "089" be considered a valid decimal floating-point value or an invalid octal integer?

Reply:
https://protobuf.dev/reference/protobuf/textformat-spec/#numeric

So it is invalid.  Rather that have this api also throw for detecting invalid, I was trying to just have this pick off if the number is a float or not, and then the actual parsing apis are what do the final validations.

Long term, I think we might want to look at bring over the C++ tokenizer concept, as it is the common base for their JSON and TextFormat parsing, so it could provide a good model to collapsing some common code between the two, and actually might make picking off some of this malformed things easier.